File: android/app/src/main/java/com/zhiligit/voicefirstdrivingassistant/model/ActionPlan.kt
package com.zhiligit.voicefirstdrivingassistant.model

data class ActionPlan(
    val summary: String,
    val requiresConfirmation: Boolean,
    val actions: List<PlannedAction>
)

sealed interface PlannedAction {
    val type: String

    data class CreateNote(
        val title: String,
        val content: String
    ) : PlannedAction {
        override val type: String = "CREATE_NOTE"
    }

    data class CreateReminder(
        val title: String,
        val scheduledAt: String
    ) : PlannedAction {
        override val type: String = "CREATE_REMINDER"
    }

    data class NoAction(val reason: String) : PlannedAction {
        override val type: String = "NO_ACTION"
    }
}

File: backend/src/action-plan.ts
import { z } from "zod";

export const ActionSchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("CREATE_NOTE"), title: z.string().min(1), content: z.string().min(1) }),
  z.object({ type: z.literal("CREATE_REMINDER"), title: z.string().min(1), scheduled_at: z.string().datetime({ offset: true }) }),
  z.object({ type: z.literal("NO_ACTION"), reason: z.string().min(1) })
]);

export const ActionPlanSchema = z.object({
  summary: z.string().min(1),
  requires_confirmation: z.boolean(),
  actions: z.array(ActionSchema).max(5)
});

export type ActionPlan = z.infer<typeof ActionPlanSchema>;
